File: Utils/python/parse_ini.py
# -*- coding: utf-8 -*-
"""
Created on Fri Sep 30 11:05:09 2016

@author: matvii
"""

def parse_ini(inifile):
    fd=open(inifile,'r')
    while True:
        iline=fd.readline()
        if 'MinTim' in iline:
            MinTim=float(iline.split('=')[1].split(' ')[0])
            break
    while True:
        iline=fd.readline()
        if 'UseAO=' in iline:
            nAO=int(iline.split('=')[1].split(' ')[0])
            break
    Date=np.zeros(nAO)
    PixScale=np.zeros(nAO)
    AOSize=np.zeros(nAO)
    Filename=[]
    tao1='AOFile='
    tao2='Date'
    tao3='PixScale'
    tao4='AOSize'
    tao='AO'+str(1)
    while len(line)!=0:
        line=fd.readline()
        if tao in line:
            break
    for j in range(1,nAO+1):
        tao='[AO'+str(j)+']'
        taon='[AO'+str(j+1)+']'
        Date[j-1]=float('nan')
        while (taon not in line) and len(line)!=0:
            line=fd.readline()
            if len(line)>0 and line[0]=='#':
                line=fd.readline()
            if tao1 in line:
                Filename.append(line.split('=')[1].split(' ')[0])
            if tao2 in line:
                Date[j-1]=float(line.split('=')[1].split(' ')[0])
            if tao3 in line:
                PixScale[j-1]=float(line.split('=')[1].split(' ')[0])
            if tao4 in line:
                AOSize[j-1]=int(line.split('=')[1].split(' ')[0].split(',')[0])
    fd.close()
    fd=open(inifile)
    line=fd.readline()
    while len(line)!=0:
        if 'EphFile' in line:
            EphFile=line.split('=')[1].split(' ')[0]
        if 'ShapeFile' in line:
            ShapeFile=line.split('=')[1].split(' ')[0]
        if 'AnglesFile' in line:
            Anglefile=line.split('=')[1].split(' ')[0]
        line=fd.readline()
    fd.close()
    tlist,vlist=Read_Shape(ShapeFile)
    fd=open(Anglefile)
    line=fd.readline()
    angles=np.array([(90-float(line.split(' ')[0]))*pi/180,float(line.split(' ')[1])*pi/180,24*2*pi/float(line.split(' ')[2])])
    fd.close()
    cao=1.731446326742403e+02
    M=np.loadtxt(EphFile)
    E=np.zeros([nAO,3])
    E0=np.zeros([nAO,3])
    for j in range(0,nAO):
        if isnan(Date[j]):
            Date[j]=read_fits_date(Filename[j])+2400000.5
        if min(abs(M[:,0]-Date[j]))>0.01:
            sys.exit("No valid date in the ephemeris file")
        Ind=np.argmin(abs(M[:,0]-Date[j]))
        E[j,:]=np.array(M[Ind,4:8])
        E0[j,:]=np.array(M[Ind,1:4])
        Date[j]=Date[j]-np.sqrt(np.dot(E[j,:],E[j,:]))/cao-MinTim
    dist=np.zeros(nAO)
    km2arcsec=np.zeros(nAO)
    up=np.zeros([nAO,3])
    
    for j in range(0,nAO):
        E0[j,:]=E0[j,:]/np.sqrt(np.dot(E0[j,:],E0[j,:]))
        dist[j]=np.sqrt(np.dot(E[j,:],E[j,:]))
        E[j,:]=E[j,:]/dist[j]
        km2arcsec[j]=1/(dist[j]*149597871.0)*180/pi*3600.0
        up[j,:]=calculate_rotated_frame(Filename[j],E[j,:])

                
            
    
    
    
    
            

File: Utils/matlab/Parse_Ini.m
function [tlist,vlist,E,E0,up,Date,angles,PixScale,AOSize,km2arcsec]=Parse_Ini(inifile)
%Parse ini file
fd=fopen(inifile);
while ~feof(fd)
    line=fgetl(fd);
    line=strtrim(line);
       if strfind(line,'MinTim')==1
        break
    end
end
MinTim=sscanf(line,'MinTim=%f');

while ~feof(fd)
    line=fgetl(fd);
       if strfind(line,'UseAO=')==1
        break
    end
end

nAO=sscanf(line,'UseAO=%d');
Date=NaN(1,nAO);
Filename=cell(1,nAO);
PixScale=0.009942*ones(1,nAO);
AOSize=150*ones(1,nAO);
tao1='AOFile';
tao2='Date';
tao3='Pixscale';
tao4='AOSize';
tao=strcat('[AO','1',']');
while ~feof(fd)
    line=fgetl(fd);
       if strfind(line,tao)==1
        break
    end
    end
for j=1:nAO
    tao=strcat('[AO',int2str(j),']');
    taon=strcat('[AO',int2str(j+1),']');
    
    
    
    while  isempty(strfind(line,taon)==1)
    line=fgetl(fd);
    line=strtrim(line);
    if ~isempty(line) && line(1)=='#'
        line=fgetl(fd);
        line=strtrim(line);
    end
       if strfind(line,tao1)==1
            Filename{j}=sscanf(line,'AOFile=%s');
       end
        if strfind(line,tao2)==1
            Date(j)=sscanf(line,'Date=%f');
        end
        if strfind(line,tao3)==1
            PixScale(j)=sscanf(line,'PixScale=%f');
        end
        if strfind(line,tao4)==1
            AOSize(j)=sscanf(line,'AOSize=%f');
        end
        if feof(fd)
            break;
        end
    end
end
fclose(fd);
fd=fopen(inifile);
while ~feof(fd)
    line=fgetl(fd);
    line=strtrim(line);
       if strfind(line,'EphFile')==1
        EphFile=sscanf(line,'EphFile=%s'); 
       end
    if strfind(line,'ShapeFile')==1
        ShapeFile=sscanf(line,'ShapeFile=%s');
    end
    if strfind(line,'AnglesFile')==1
        Anglefile=sscanf(line,'AnglesFile=%s');
    end
end


fd2=fopen(Anglefile);
Angles=fscanf(fd2,'%f %f %f');
fclose(fd2);
angles=[deg2rad(90-Angles(1)) deg2rad(Angles(2)) 2*pi/Angles(3)*24];
fclose(fd);
%Read ephm information
cao=1.731446326742403e+02;
M=dlmread(EphFile);
E0=[];
E=[];
for j=1:size(M,1)
    %If date is not set in ini file, read it from the fits file
    if isnan(Date(j))
        Date(j)=read_fits_date(Filename{j})+2400000.5;
    end
    [I,J]=min(abs(M(:,1)-Date(j)));
    if I>1e-2
        error('MJD-OBS time not found in ephm.dat')
    end
    E0(j,:)=M(J,2:4);
    E(j,:)=M(J,5:7);
    Date(j)=Date(j)-norm(E(j,:))/cao-MinTim;
end
dist=[];
km2arcsec=[];
up=zeros(j,3);
%Scaling and rot information 
for j=1:length(Date)
    E0(j,:)=E0(j,:)/norm(E0(j,:));
    dist(j)=norm(E(j,:));
    E(j,:)=E(j,:)/dist(j);
    km2arcsec(j)=1/(dist(j)*149597871)*180/pi*3600;
    up(j,:)=calculate_rotated_frame(Filename{j},E(j,:));
end
[tlist,vlist]=read_shape(ShapeFile,1);
end
